cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,504 @@
+    {
+        value_squared = value*value;
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMeanVar(T _value, T _value_squared=0, gdf_index_type _count=0)
+    : value(_value), value_squared(_value_squared), count(_count)
+    {};
+
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMeanVar()
+    : value(0), value_squared(0), count(0)
+    {};
+
+    using this_t = ColumnOutputMeanVar<T, true>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared),
+            (this->count + rhs.count)
+        );
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared) &&
+            (this->count == rhs.count)
+        );
+    }
+};
+
+// @overload ColumnOutputMeanVar<T, false>
+template<typename T>
+struct ColumnOutputMeanVar<T, false>
+{
+    T value;
+    T value_squared;
+    gdf_index_type count;
+
+    template<typename T_input>
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMeanVar(T_input _value, bool is_valid)
+    : value( static_cast<T>(_value) )
+    {
+        value_squared = value*value;
+    };
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMeanVar(T _value, T _value_squared=0, gdf_index_type _count=0)
+    : value(_value), value_squared(_value_squared), count(_count)
+    {};
+
+    CUDA_HOST_DEVICE_CALLABLE
+    ColumnOutputMeanVar()
+    : value(0), value_squared(0), count(0)
+    {};
+
+    using this_t = ColumnOutputMeanVar<T, false>;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    this_t operator+(this_t const &rhs) const
+    {
+        // count won't be updated.
+        return this_t(
+            (this->value + rhs.value),
+            (this->value_squared + rhs.value_squared),
+            (this->count)
+        );
+    }
+
+    CUDA_HOST_DEVICE_CALLABLE
+    bool operator==(this_t const &rhs) const
+    {
+        // count won't be compared.
+        return (
+            (this->value == rhs.value) &&
+            (this->value_squared == rhs.value_squared)
+        );
+    }
+};
+
+// --------------------------------------------------------------------------------------------------------
+/** -------------------------------------------------------------------------*
+ * @brief input struct with/without null bitmask
+ * A helper struct struct for `column_input_iterator`
+ * `ColumnInput.at(gdf_index_type id)` computes
+ * `data` value and valid flag at `id` and construct and return
+ * `T_output(data, valid flag)`.
+ * If nulls_present = true, the valid flag is always true.
+ * If nulls_present = false, the valid flag corresponds to null bitmask flag
+ * at `id`, and the data value = (is_valid(id))? data[id] : identity;
+ *
+ * @tparam  T_output a scalar data type to be output
```

Comment:
This needs a clearer name than `T_output`. Furthermore, it is not a scalar data type. It's some type that can be constructed with arguments `(T, bool)`. 

Likewise, it needs clearer documentation on what this object does. In reality, it's more of a "Loader" or "Mutator" object, that given some value, potentially mutates it into a new value. 

Finally, and most importantly, does `ColumnInput` really need to be complicated with this mutator object? From a design perspective, I believe `ColumnInput` should be simplified such that it does only one thing: for a requested value at index `i`, if the value is valid, return it. Else, return an `identity` value. 

Any more complex functionality using your mutator objects should be built _on top of_ this simplified `ColumnInput`. Right now, they are intertwined and I think it makes the implementation harder to follow.